src/lib.rs

```diff
@@ -239,120 +238,132 @@ impl<'text> InitialInfoExt<'text> {
         text: &'a str,
         default_para_level: Option<Level>,
     ) -> InitialInfoExt<'a> {
-        let mut original_classes = Vec::with_capacity(text.len());
+        let (original_classes, paragraphs, pure_ltr) =
+            compute_initial_info(data_source, text, default_para_level);
 
-        // The stack contains the starting byte index for each nested isolate we're inside.
-        let mut isolate_stack = Vec::new();
-        let mut paragraphs = Vec::new();
-        let mut pure_ltr = Vec::new();
+        InitialInfoExt {
+            base: InitialInfo {
+                text,
+                original_classes,
+                paragraphs,
+            },
+            pure_ltr,
+        }
+    }
+}
+
+/// Implementation of initial-info computation.
+fn compute_initial_info<D: BidiDataSource, T: TextSource + ?Sized>(
```

Comment:
assuming that the diffs to this function in commit 3df9426863f72482fd7a4ee36e44692c6ad6a084 are idempotent. looks like some match arms were moved around leading to a larger diff than this actually is, and I don't see any semantic changes.

Please let me know if my read here is incorrect!